app/src/store.js

```diff
@@ -0,0 +1,243 @@
+        },
+
+        updateRecords(state, records) {
+            /*
+             * Set the array of currently-displayed Records.
+             *
+             * @param {array} records - An array of Records, formatted as JSON
+             *                          blobs, typically returned from the Grout
+             *                          API.
+             */
+            state.records = records;
+        },
+
+        updateFilter(state, payload) {
+            /*
+             * Update the active filters based on user input.
+             *
+             * @param {Object} payload - Data to update the filter.
+             * @param {string} payload.schemaName - The name of the schema (the base
+             *                                      key in the `state.filters` object
+                     *                              in which to store this data).
+             * @param {string} payload.fieldName - The name of the field that this query
+             *                                     pertains to.
+             * @param {string} payload.fieldType - The type of field represented by this
+             *                                     filter (one of 'text', 'select',
+             *                                     'min', or 'max').
+             * @param {string} payload.query - The user-inputted query value.
+             */
+            // Validate input parameters.
+            const schemaName = payload.schemaName;
+            const fieldName = payload.fieldName;
+            const fieldType = payload.fieldType;
+            const query = payload.query;
+
+            const paramDict = {
+                'schemaName': schemaName,
+                'fieldName': fieldName,
+                'fieldType': fieldType,
+                'query': query
+            };
+
+            Object.keys(paramDict).forEach(paramName => {
+                let param = paramDict[paramName];
+                if (!param) {
+                    throw new Error(`
+                        Function argument "${paramName}" is required in order
+                        to update filters.
+                    `)
+                }
+            });
+
+            // Check whether an entry already exists for this schema. If not,
+            // initialize it.
+            if (!state.filters[schemaName]) {
+                state.filters[schemaName] = {};
+            }
+
+            // If this is a number query (or another type of field
+            // that accepts multiple inputs, like a multiselect)
+            // the nested query object may already exist, so check
+            // for it before creating it.
+            let nestedQuery = {};
+            if (!state.filters[schemaName][fieldName]) {
+                state.filters[schemaName][fieldName] = {};
+            } else {
+                nestedQuery = state.filters[schemaName][fieldName];
+            }
+
+            // Construct the actual query.
+            switch(fieldType) {
```

Comment:
This query construction logic feels to me like it should happen somewhere in the Grout API, since its role is essentially packaging filters for the specific format that the API expects -- a Grout API user shouldn't really have to care about that. But I'm not entirely sure where it should go! My current idea is that maybe the current implementation of `Schema.getFilters` should be called something like `Schema.getFilterables`, and then most of this logic could be encapsulated in a method called `Schema.getFilter(filterable, query)` that takes a filterable field and a query and formats it for the Grout API.

Reply:
Thinking about this more, I wonder if the `Record` might not be a better place for it? I was imagining my ideal API for this, and I think I would want to be able to call something kind of Django-like, for example `Record.query(schema, [filter1, filter2, filter3])`. Each `filter` object could be some sort of simplified specification of the filtering that I want to take place, something like
```
{
form: 'details',
field: 'name',
filterValue: 'Bob',
filterType: 'contains'
}
```
That's a bit of a sketch, but the idea was that after using `Schema.getFilters` to generate filter elements, each filter element would output one of those objects, and then the list of those objects could get passed into the `Record` query function. Does that make sense? I'm just sort of spitballing here so feel free to ignore. 🙂

Reply:
I like that a lot! (It seems like this comment is related to your comment above about returning an iterable from that method, so I'm going to address them both here.) As a developer I would definitely expect a method like `getFoos` to return an iterable data type, so I think the idea of returning an array of objects from `Schema.getFilters` is really smart. Passing filters as optional args to `Record.query` instead of as a massive object feels more elegant to me, too.

One disadvantage of returning an array for `Schema.getFilters` is that it'd be harder to group filters by form, which seems like a pretty common use case. I can see a couple of approaches to this:

1. Do nothing; if the user wants a hash map, they can build one out of the arrays
2. Provide a convenience function, something like `Schema.getFiltersByForm(formName)`, for getting a hash map
3. Allow an optional `formName` parameter for `getFilters`, e.g. `Schema.getFilters(formName)`, that restricts it to just one form name (with an additional method like `Schema.getForms()` that returns the form names)
4. Return nested arrays, one for each form, e.g.:

```
[
  [
    {
      form: 'details',
      field: 'name',
      filterValue: 'Bob',
      filterType:  'contains'
    },
    {
      form: 'details',
      field: '
  ],
  [
    {
      form: 'misc',
      field: 'foo',
      filterValue: 'Alice',
      filterType: 'containment_multiple'
    }
  ]
]
```

None of these seems like a _great_ solution to me, but 3 seems like perhaps the prettiest solution.

Reply:
It looks like [`_.groupBy()`](https://lodash.com/docs/4.17.10#groupBy) does pretty much exactly this, so I don't think it would be too problematic to just leave that up to the user if they want it. But it may be common enough that we may still want to make it easier for people, and in that case, of the other options, I think I agree that #3 seems the most usable.